### PR TITLE
Pull request for #3: Stata package installation

### PR DESCRIPTION
**Pull request by M-R-Sullivan**
For #3. I used [this repo](https://github.com/sergiocorreia/stata-misc) as a model for allowing installation directly from GitHub via the Stata console. 

You should be able to install the repo's packages by following the readme instructions, although you'll need to substitute `issue3-install_packages` for `master` in the paths to GitHub.

Let me know if you think more extensive attribution/licensing is warranted.

### CODE REVIEW

File: crosswalk/ado/stata.toc [hunk 1]
```diff
@@ -0,0 +1,3 @@
+p make_input_template Write an input template
```
Comment (arosenbe) [inline crosswalk/ado/stata.toc]: Did Matt and Jesse write this? If so, we should give them attribution as for xtreg_cg.

File: gslab_misc/ado/stata.toc [hunk 2]
```diff
@@ -0,0 +1,34 @@
+p aivreg                Estimates panel IV models using 2SLS within estimator
```
Comment (arosenbe) [inline gslab_misc/ado/stata.toc]: Maybe mention that this is a collection of third part packages and provide attribution in the various .pkg files.

File: xtreg_ch/ado/xtreg_ch_loglik.pkg [hunk 3]
```diff
@@ -0,0 +1,6 @@
+d xtreg_ch
+d
+d xtreg_ch_loglik: log likelihood for xtreg_ch
+d
+f xtreg_ch_loglik.ado
```
Comment (arosenbe) [inline xtreg_ch/ado/xtreg_ch_loglik.pkg]: No help?

File: xtreg_ch/readme.md [hunk 4]
```diff
@@ -1,2 +1,13 @@
-xtreg_ch runs a random-effects regression allowing for parametric correlation between RHS variables and random effect as well as heteroskedasticity in individual errors.
+xtreg_ch runs a random-effects regression allowing for parametric correlation 
+between RHS variables and random effect as well as heteroskedasticity in 
+individual errors.
```
Comment (arosenbe) [inline xtreg_ch/readme.md]: Why the line breaks?

Comment (M-R-Sullivan) [inline xtreg_ch/readme.md]: Ease of reading.

File: xtreg_ch/ado/xtreg_ch_loglik.pkg (hunk 3 above)
Comment (M-R-Sullivan) [inline xtreg_ch/ado/xtreg_ch_loglik.pkg]: There wasn't one before.

File: crosswalk/ado/stata.toc (hunk 1 above)
Comment (M-R-Sullivan) [inline crosswalk/ado/stata.toc]: It wasn't clear to me whether they did or not.

File: gslab_misc/ado/cf_mg.pkg [hunk 5]
```diff
@@ -0,0 +1,6 @@
+d gslab_stata
+d
+d cf_mg: Like -cf-, but the exit flag is always zero
+d
+f cf_mg.ado
```
Comment (M-R-Sullivan) [inline gslab_misc/ado/cf_mg.pkg]: There wasn't a hlp file in the repo before.

File: gslab_misc/ado/plotcoeffs_nolab.pkg [hunk 6]
```diff
@@ -0,0 +1,6 @@
+d gslab_stata
+d
+d plotcoeffs_nolab: Plot coefficients after a regression (no labels)
+d
+f plotcoeffs_nolab.ado
```
Comment (M-R-Sullivan) [inline gslab_misc/ado/plotcoeffs_nolab.pkg]: There wasn't a help file in the repo, I believe.

File: gslab_misc/ado/cutby.pkg [hunk 7]
```diff
@@ -0,0 +1,7 @@
+d gslab_stata
+d
+d cutby: By-able version of egen ... = cut()
+d
+f cutby.hlp
```
Comment (M-R-Sullivan) [inline gslab_misc/ado/cutby.pkg]: Good catch!

**Comment by M-R-Sullivan** [comment]
Note to self: get all RAs to reinstall `gslab_stata` once this passes PR. 

📊

File: gslab_misc/ado/checkdta.pkg [hunk 8]
```diff
@@ -0,0 +1,8 @@
+d gslab_stata
+d
+d checkdta: Create checkdta.log, which allows users to check whether Stata
+datasets have actually changed after make.bat is re-run.
```
Comment (M-R-Sullivan) [inline gslab_misc/ado/checkdta.pkg]: No

File: gslab_misc/ado/cf_mg.pkg [hunk 9]
```diff
@@ -0,0 +1,6 @@
+d gslab_stata
+d
+d cf_mg: Like -cf-, but the exit flag is always zero
```
Comment (M-R-Sullivan) [inline gslab_misc/ado/cf_mg.pkg]: I will keep as is.

File: gslab_misc/ado/load_and_append.pkg [hunk 10]
```diff
@@ -0,0 +1,7 @@
+d gslab_stata
+d
+d load_and_append: Load_and_append appends a group of files and loads the appended file into memory.
```
Comment (M-R-Sullivan) [inline gslab_misc/ado/load_and_append.pkg]: I will standardize

File: gslab_misc/ado/placevar.pkg [hunk 11]
```diff
@@ -0,0 +1,6 @@
+d gslab_stata
+d
+d placevar: Orders varlist
+f placevar.ado
```
Comment (M-R-Sullivan) [inline gslab_misc/ado/placevar.pkg]: Fixed

File: gslab_misc/ado/rankunique.pkg [hunk 12]
```diff
@@ -0,0 +1,7 @@
+d gslab_stata
+d
+d rankunique: Replacement for egen ... = rank(), unique that uses explicit randomization
```
Comment (M-R-Sullivan) [inline gslab_misc/ado/rankunique.pkg]: OK.

File: gslab_misc/ado/stata.toc (hunk 2 above)
Comment (M-R-Sullivan) [inline gslab_misc/ado/stata.toc]: - I added author attributions
- All of the ado files except two either had no authorship label or had help files lising MG, JMS, and/or RAs as authors. Those two are:
    - matrix_to_txt.ado: by  Michael Blasnik (M. Blasnik & Associates, Boston, MA), Ben Jann (ETH Zurich), and Matthew Gentzkow
    - placevar.ado: by Jeroen Weesie (Dept of Sociology/ICS, Utrecht University)

**Comment by M-R-Sullivan** [comment]
Still working on this.

**Comment by M-R-Sullivan** [comment]
@gentzkow and @jmshapir:

We noticed the following things when preparing this repo for its public release.
- I thought @arosenbe's suggestion that we give this repo an MIT license was a good one, but I noticed that two scripts in `gslab_misc` have authors other than either of you or former RAs. Namely, (i)  Michael Blasnik (M. Blasnik & Associates, Boston, MA) and Ben Jann (ETH Zurich) are listed among the authors of `matrix_to_txt`, and (ii) Jeroen Weesie (Dept of Sociology/ICS, Utrecht University) is the author of `placevar`. If you don't think we should put these files in the licensed repo, we could remove them or reconsider making the repo public. Also,  there are some other ado files without listed authors.

- A few of the unit tests for this repo's ado files fail. Do you think it is worth spending time to understand why and implement fixes as a part of this pull request, or would you like to proceed in some other way?

**Comment by jmshapir** [comment]
@M-R-Sullivan we should see what @gentzkow thinks but here are my reactions:
* If those ADO files are literally from the outside world, then they probably shouldn't be in ```gslab-misc``` anyway.  If we want to install third-party tools as part of our setup, we should probably use [ssc](http://www.stata.com/support/ssc-installation/). Certainly we do not want to try to license anything we did not write.
* I would venture it's worth an hour or so of work to debug unit tests, not more. But we can see what @gentzkow thinks. I am happy to defer to him.

**Comment by M-R-Sullivan** [comment]
@arosenbe - thanks for your helpful comments! I acted on most of your suggestions. There are a few things left to do that I was hoping you could wrap up. Some notes:
- I added an SVN-free testing script to `gslab_misc`.  If you dislike my approach, you can change it here or in a follow-up issue. If you like it, you can add it to the other packages here or in a follow-up issue.
- I added author attributions but not versions to the package files. You/another lab member can add them if you think they'd provide some benefit.
- Per our Skype discussion, the provenance of some of the `gslab_misc` packages is unclear. It is probably best to remove all of those that we aren't sure we wrote before adding a license. I initially made this issue because we needed a way of getting some (maybe just `matrix_to_txt`) of these ado files for `divergence`. So you should probably make a note to update the `divergence` documentation with the new instructions for installing the required ado files.

Let me know if I can help out with anything here!

**Comment by lboxell** [comment]
@M-R-Sullivan,

Quick question: Did you ever successfully `net install` from the current branch?

**Comment by M-R-Sullivan** [comment]
Indeed I did!

**Comment by lboxell** [comment]
@arosenbe,

The net installation works (I needed to delete the `/tree` part of the URL), and I've removed the packages that MG didn't confirm we created (with the note regarding mat2txt included). 

I haven't included a license at the top level, as I'm not sure what the best way to handle this is with the matrix_to_txt function. 

I'm inclined to make a decision regarding the license and wrap this up, since the primary purpose was to just get proper download instructions.

**Comment by arosenbe** [comment]
@lboxell, I'd read [this summary by Nick Cox](https://www.stata.com/statalist/archive/2004-03/msg00840.html), add a license at the top level with plenty of acknowledgement for mat2txt's original authors (noting specifically what they did and did not do), and call it a day.

**Comment by lboxell** [comment]
@arosenbe,

See my commit above for license/readme changes. Let me know if you have any other comments or if I can go ahead and squash and merge.

**Comment by arosenbe** [comment]
Good to go @lboxell. it's about time we wrapped this up.

**Pull request merged**